lec05-dplyr.Rmd

```diff
@@ -1,4 +1,251 @@
+> ### Challenge
+>
+> 1. Make a wide data frame with `year` as columns, `plot_id` as rows, and the
+>   values are the number of genera per plot. You will need to summarize before
+>   reshaping, and use the function `n_distinct` to get the number of unique
+>   types of a genera. It's a powerful function! See `?n_distinct` for more.
+>
+> 2. Now take that data frame, and make it long again, so each row is a unique
+>    `plot_id` `year` combination.
+>
+> 3. The `surveys` data set is not truly wide or long because there are
+>    two columns of measurement - `hindfoot_length` and `weight`.  This makes it
+>    difficult to do things like look at the relationship between mean values of
+>    each measurement per year in different plot types. Let's walk through a
+>    common solution for this type of problem. First, use `gather` to create a
+>    truly long dataset where we have a key column called `measurement` and a
+>    `value` column that takes on the value of either `hindfoot_length` or
+>    `weight`. Hint: You'll need to specify which columns are being gathered.
+>
+> 4. With this new truly long data set, calculate the average of each
+>    `measurement` in each `year` for each different `plot_type`. Then
+>    `spread` them into a wide data set with a column for `hindfoot_length` and
+>    `weight`. Hint: Remember, you only need to specify the key and value
+>    columns for `spread`.
+
+<!---
+  ```{r, echo=FALSE, purl=FALSE}
+## Answer 1
+rich_time <- surveys %>%
+  group_by(plot_id, year) %>%
+  summarize(n_genera = n_distinct(genus)) %>%
+  spread(year, n_genera)
+
+head(rich_time)
+
+## Answer 2
+rich_time %>%
+  gather(year, n_genera, -plot_id)
+
+## Answer 3
+surveys_long <- surveys %>%
+  gather(measurement, value, hindfoot_length, weight)
+
+## Answer 4
+surveys_long %>%
+  group_by(year, measurement, plot_type) %>%
+  summarize(mean_value = mean(value, na.rm=TRUE)) %>%
+  spread(measurement, mean_value)
+```
+--->
+
+## Exporting data
+
+Now that you have learned how to use **`dplyr`** to extract information from
+or summarize your raw data, you may want to export these new datasets to share
+them with your collaborators or for archival.
+
+Similar to the `read_csv()` function used for reading CSV files into R, there is
+a `write_csv()` function that generates CSV files from data frames.
+
+Before using `write_csv()`, we are going to create a new folder, `data_output`,
+in our working directory that will store this generated dataset. We don't want
+to write generated datasets in the same directory as our raw data. It's good
+practice to keep them separate. The `data` folder should only contain the raw,
+unaltered data, and should be left alone to make sure we don't delete or modify
+it. In contrast, our R code will generate the contents of the `data_output`
+directory, so even if the files it contains are deleted, we can always
+re-generate them.
+
+Use the `getwd()` function to find out which is the current working directory.
```

Comment:
This may or may not start to get confusing. File systems for some reason are a challenge to learn for beginners, especially for Windows users.

This is also not super important once they start using R projects. But still, good to cover I think :)